src/userspace.poc.zig

```diff
@@ -150,117 +125,123 @@ fn test_segfault() linksection(".userspace") void {
-            putstr("process ");
-            putnbr(wait_ret);
-            switch (status.type) {
-                .Exited => {
-                    putstr(" exited with exit code ");
-                    putnbr(status.value);
-                },
-                .Continued => putstr(" continued"),
-                .Stopped => putstr(" stopped"),
-                .Signaled => {
-                    putstr(" received signal ");
-                    putnbr(status.value);
-                },
-            }
-            putchar('\n');
-        }
-
-        _ = syscall(.exit, .{0});
+var byte: u8 linksection(".userspace") = 0;
+var byte_index: u5 linksection(".userspace") = 0;
+
+fn server_handler(id: u32) linksection(".userspace") callconv(.C) void {
+    // putstr("bonjour\n");
+    byte |= @truncate((id - @intFromEnum(signal.Id.SIGUSR1)) << byte_index);
+    byte_index += 1;
+    if (byte_index == 8) {
+        _ = syscall(.write, &.{ &byte, 1 });
+        byte_index = 0;
+        byte = 0;
     }
+}
 
-    const sigaction = @import("task/signal.zig").Sigaction{ .sa_sigaction = &poc_sigaction, .sa_flags = .{
-        .SA_SIGINFO = true,
-    } };
-    _ = syscall(.sigaction, &.{ 2, &sigaction, null });
-    // _ = syscall(.signal, &.{ 2, &poc_signal });
-    _ = syscall(.kill, &.{ syscall(.getpid, &.{}), 2 });
-    {
-        const pid = syscall(.fork, .{});
-        if (pid == 0) {
-            server();
-        } else {
-            client(@bitCast(pid));
-        }
+fn server() linksection(".userspace") void {
+    _ = syscall(.signal, &.{ @intFromEnum(signal.Id.SIGUSR1), &server_handler });
+    _ = syscall(.signal, &.{ @intFromEnum(signal.Id.SIGUSR2), &server_handler });
+    while (true) {}
+}
+
+fn send_byte(pid: u32, c: u8) linksection(".userspace") void {
+    for (0..8) |index| {
+        _ = syscall(.kill, &.{ pid, @intFromEnum(signal.Id.SIGUSR1) + ((c >> @as(u3, @intCast(index))) & 1) });
+        _ = syscall(.sleep, .{10});
     }
+}
 
-    _ = syscall(.exit, .{0});
+fn send(pid: u32, msg: []const u8) linksection(".userspace") void {
+    for (msg) |c| {
+        send_byte(pid, c);
+    }
+}
 
-    // while (true) {}
-    _ = syscall(.write, &.{ "coucou\n", 7 });
-    _ = syscall(.fork, .{});
+fn client(server_pid: u32) linksection(".userspace") void {
     _ = syscall(.sleep, .{100});
-    _ = syscall(.write, &.{ "bonjour\n", 8 });
-    _ = syscall(.fork, .{});
+    send(server_pid, "le minitalk\n");
     _ = syscall(.sleep, .{100});
-    _ = syscall(.write, &.{ "salut\n", 6 });
-    _ = syscall(.exit, .{42});
 }
 
-export fn _userland() linksection(".userspace") void {
-    zig_userland();
+export fn userland_minitalk() linksection(".userspace") void {
+    const pid = syscall(.fork, .{});
+    if (pid == 0) {
+        server();
+    } else {
+        client(@bitCast(pid));
+    }
+    _ = syscall(.exit, .{0});
+}
+
+export fn userland_sleep() linksection(".userspace") void {
+    _ = syscall(.sleep, .{100000});
```

Comment:
Maybe add an exit.
Moreover, maybe sleep(20000) is a bit long for a demo

```suggestion
    _ = syscall(.sleep, .{20000});
    _ = syscall(.exit, .{0});
```